src/engraving/rendering/dev/tdraw.cpp

```diff
@@ -2288,6 +2288,30 @@ void TDraw::draw(const Note* item, Painter* painter)
                 painter->setPen(item->selected() ? config->criticalSelectedColor() : config->criticalColor());
             }
         }
+
+        if (item->chord() && item->chord()->segment() && item->staff()
+            && !item->score()->printing() && MScore::warnChordsPitchRange && !item->staff()->isDrumStaff(item->chord()->tick())) {
+            const Instrument* in = item->part()->instrument(item->chord()->tick());
+            if (in->family() == "keyboards") {
+                int i = item->ppitch();
+                if (item->chord()->notes().size() >= in->maxNotesPerChordsP()
+                    || (item->chord()->withUpNoteDistance(i) > 24)
+                    || (item->chord()->withDownNoteDistance(i) > 24)) {
+                    item->chord()->setIsImpossibleChord(true);
+                    item->chord()->setColor(item->selected() ? config->criticalSelectedColor() : config->criticalColor());
```

Comment:
Calling `chord()->setColor()` is unsafe, because it modifies the property of the chord itself but doesn't register that change in the undo stack. During drawing, all elements must be treated as read-only, so you shouldn't change properties of them. To still change the color, you can either change the color of the pen of the painter, or make changes to the `curColor` method.